client/src/main/java/org/eclipse/hono/client/impl/ProtocolAdapterCommandConsumerFactoryImpl.java

```diff
@@ -196,19 +198,29 @@ protected void onDisconnect() {
     }
 
     private Future<Void> removeCommandConsumer(final String tenantId, final String deviceId,
-            final SpanContext onCloseSpanContext) {
+            final SpanContext onCloseSpanContext, final boolean isUnlimitedLifespanConsumer) {
         log.trace("remove command consumer [tenant-id: {}, device-id: {}]", tenantId, deviceId);
         adapterInstanceCommandHandler.removeDeviceSpecificCommandHandler(tenantId, deviceId);
 
         return deviceConnectionClientFactory.getOrCreateDeviceConnectionClient(tenantId)
                 .compose(client -> client.removeCommandHandlingAdapterInstance(deviceId, adapterInstanceId,
                         onCloseSpanContext))
+                .compose(removed -> {
+                    if (!removed && isUnlimitedLifespanConsumer) {
```

Comment:
I find it extremely confusing that *DeviceConnectionClientImpl.removeCommandHandlingAdapterInstance* method does not comply with the Device Connection API in that the returned future actually succeeds for status codes 404 and 412 but fails for e.g. a status code 404 being returned from the service.
I am even more confused seeing that a few lines below we actually map the successful result to a 404?

Reply:
The change to let *DeviceConnectionClientImpl.removeCommandHandlingAdapterInstance* return a `Boolean` future value was done when introducing the lifespan for the corresponding cache entries, with which 404 exceptions when trying to remove a just expired entry can easily occur (when the ttd has elapsed). PR for that was #1926. With that change, a failure to remove a (potentially expired) entry will also not be traced as an error in the client anymore.
I think the corresponding server method wasn't changed back then for compatibility reasons and following up on this later on was abandoned because of the new command router concept.
I see that it may be confusing to use that client method with having the API definition in mind. I've changed the javadoc now to add a note about that difference.
We could also revert #1926 and just adapt the client tracing to skip logging a 412 error there. If doing so, that should be in a follow-up PR, but frankly, I don't see that much benefit in that, also since the plan is to abandon the device connection API for the command routing component concept.

> I am even more confused seeing that a few lines below we actually map the successful result to a 404?

I've changed that now to 412.